plugins/instanceof-checkcast/src/main/java/org/qbicc/plugin/instanceofcheckcast/ClassInitializerRegister.java

```diff
@@ -0,0 +1,54 @@
+package org.qbicc.plugin.instanceofcheckcast;
+
+import java.util.function.Consumer;
+
+import org.qbicc.context.ClassContext;
+import org.qbicc.context.CompilationContext;
+import org.qbicc.plugin.reachability.RTAInfo;
+import org.qbicc.type.definition.DefinedTypeDefinition;
+import org.qbicc.type.definition.LoadedTypeDefinition;
+import org.qbicc.type.definition.element.InitializerElement;
+
+/**
+ * Register all reachable class's class initalizer methods as entrypoints
+ * so they survive to be lowered into functions.
+ * This is required as we lose info after ReachabilityBlockBuilder required
+ * to otherwise keep them alive.
+ * 
+ * Eventually, this will need to skip an class initializer that was run as
+ * part of the build process.
+ */
+public class ClassInitializerRegister implements Consumer<CompilationContext>  {
+
+    @Override
+    public void accept(CompilationContext ctxt) {
+        RTAInfo rtaInfo = RTAInfo.get(ctxt);
+		
+        ClassContext classContext = ctxt.getBootstrapClassContext();
+        DefinedTypeDefinition jloDef = classContext.findDefinedType("java/lang/Object");
+        LoadedTypeDefinition jlo = jloDef.load();
+
+        // Code below uses #hasMethodBody() to filter out the empty
+        // class initializers and to avoid processing internal arrays
+
+        // Register Object's <clinit> if it exists
+        InitializerElement object_clinit = jlo.getInitializer();
+        if (object_clinit != null && object_clinit.hasMethodBody()) {
+            ctxt.registerEntryPoint(object_clinit);
```

Comment:
Are you suggesting re-enqueueing every live classes initializer during reachability analysis?  I had originally tried to be more selective but we lose the information needed as we lower further but it might work to re-enqueue every reachable class's initializer.

Longer term, I'd like to add a `InitializeClass` node which could act as an anchor for reachability of class initializers if they get added eagerly.  Couple this with an optimization pass that removes the `InitializeClass` nodes when we can prove they've been pre-initialized or aren't needed would get us to a fairly good story.

Do you want to merge this now and then try your proposal to re-enqueue during reachability in a separate PR?

Reply:
> Are you suggesting re-enqueueing every live classes initializer during reachability analysis? I had originally tried to be more selective but we lose the information needed as we lower further but it might work to re-enqueue every reachable class's initializer.

In part.  But what's going on in the back of my mind is, I'm not sure we *want* to lower initializers like this (as-is) in the end.  We're really basing this on the assumption that the initializer must make it through to run time but this might not be the final model we even want.  And also in the back of my mind is that we really need to *at a minimum* be compiling initializers completely, if not actually running them, *before* proceeding with compiling things that would require initialization, which is something that is broken today and will certainly be more broken by enqueueing initializers *even later* than we are now.

For the sake of discussion (and this is something that we can and should discuss on the call), what if *all* initializers were *only* build time (recognizing that yes this would clearly break a lot of existing code, but just for the sake of imagining a long term model), and anything that is either initialized at run time or initialized twice (once during build and again at run time) is done using a *different* mechanism?

One such mechanism could be dynamic constants.  We could determine the nature of the constant based on modifiers (well, annotations to start) of their bootstrap methods, for example.  Since `javac` can't yet produce any such thing, we could use annotation processing and bytecode manipulation to set these up, or substitute an alternative mechanism.

An idea to get things going in the nearer term would be to statically split the class initializer into build time and run time components (by walking backwards from static field assignments - I have an algorithm in mind), where each run time component could be mapped to individual initialization processes for accesses to particular fields.

Using this model, class initializers would never have to leave the `ADD` stage, greatly simplifying things.

> Longer term, I'd like to add a `InitializeClass` node which could act as an anchor for reachability of class initializers if they get added eagerly. Couple this with an optimization pass that removes the `InitializeClass` nodes when we can prove they've been pre-initialized or aren't needed would get us to a fairly good story.

An initialization check node seems like a good way to represent this concept; we could theoretically apply this concept to dynamic constants too.

> Do you want to merge this now and then try your proposal to re-enqueue during reachability in a separate PR?

I will put a pin in that for a moment, but yeah maybe we could do that.

Reply:
I think you're hitting on a key problem here - Java doesn't have the tools today for a programmer to express their intent about build time vs runtime initialization.  All programmers have is a `static {}` block and, quite frankly, the wrong cost model based on dynamic JVMs which encourage making expensive operations as lazy as possible to delay initialization costs until first use.  With a static compiler, we need a way for users to express they want those expensive operations done at **build** time rather than runtime.  The existing trend in Java ecosystem toward lazier initialization doesn't help the static compiler approach.

> For the sake of discussion (and this is something that we can and should discuss on the call), what if _all_ initializers were _only_ build time (recognizing that yes this would clearly break a lot of existing code, but just for the sake of imagining a long term model), and anything that is either initialized at run time or initialized twice (once during build and again at run time) is done using a _different_ mechanism?

I've with you on needing new tools and new ways to look at initialization.  Also a +1 to preferring build time initialization.  But "initialized twice" is a very concerning model to me without a both the compiler tracking such values to ensure that no computed constants based on the compile-time value make it into the static image, and some way for the application developer to indicate what proper re-initialization is.

> An idea to get things going in the nearer term would be to statically split the class initializer into build time and run time components (by walking backwards from static field assignments - I have an algorithm in mind), where each run time component could be mapped to individual initialization processes for accesses to particular fields.

I'm very interested by this and looking forward to seeing the algorithm.  One of the non-obvious challenges is how to handle failure to initialize one of these fields.  Today, there's a state machine driven by class initialization that flags a class as unusable if a `static {}` block fails.  With per-field initializers, there's an odd state where a class may be in use before some of these field-specific initializers have run (and failed!).

Thinking further out, JNI use may require that all field initializers for a class are run when passed to a JNI method or when JNI does field / method lookup on it.

Lots of details to figure out here!  Looking forward to the discussion